In [1]:
import pandas as pd
import numpy as np
from IPython.display import display

## Basic Data Frame Examination

<font size="3">Start off with a basic examination of the dataframes using `Pandas` standard EDA functions.</font>

In [2]:
species = pd.read_csv('data/species_info.csv')
observations = pd.read_csv('data/observations.csv')

In [3]:
species.head(3)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN

In [4]:
observations.head(3)

scientific_name                            park_name  observations
0  Vicia benghalensis  Great Smoky Mountains National Park            68
1      Neovison vison  Great Smoky Mountains National Park            77
2   Prunus subcordata               Yosemite National Park           138

In [5]:
display(species.shape)
observations.shape

(5824, 4)

(23296, 3)

In [6]:
species.info()
print()
observations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23296 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23296 non-null  object
 1   park_name        23296 non-null  object
 2   observations     23296 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 546.1+ KB


<font size="3">There are only null values in the species dataframe are under the column conservation status. Some more familiarization of the species data set would be helpful before making decision on the null values.</font>
## EDA and Data Cleaning
    
#### Species Data Frame Analysis
    
<font size="3">Let's move on to further investigate the data and clean it up as much as possible. I will start with a look at duplicate entries.</font>

In [7]:
# species.conservation_status.value_counts()
con_status_counts = species.conservation_status.value_counts(dropna=False)
con_status_counts

NaN                   5633
Species of Concern     161
Endangered              16
Threatened              10
In Recovery              4
Name: conservation_status, dtype: int64

In [8]:
display(con_status_counts[1:].sum()/species.shape[0])
con_status_counts[0]/species.shape[0]

0.03279532967032967

0.9672046703296703

<font size="3">Only 3% of the data set contains various conservation statuses while 97% are NaN values. My initial guess is that these species are `not in danger` and are experiencing healthy numbers.<br><br>Let's do a little more investigation before making this conclusion.</font>

In [9]:
species.category.value_counts()

Vascular Plant       4470
Bird                  521
Nonvascular Plant     333
Mammal                214
Fish                  127
Amphibian              80
Reptile                79
Name: category, dtype: int64

<font size="3">Plants make up the vast majority of the species while the rest of the species are animals from various classes.</font>

In [10]:
display(species.scientific_name.nunique())
display(observations.scientific_name.nunique()) 
species.common_names.nunique()

5541

5541

5504

<font size="3">The dataframe observations only contains the scientific name and not the common name. There are 37 less total number of common names compared to scientific names, indicating that some species may share a common name. There are 5824 total rows in the data frame indicating 283 possible duplicates in the species scientific names.<br><br>The observations data frame also contains 5541 unique scientific names. To make sure all species are included in both, a check is performed to see if the unique names from both data sets are equal.</font>

In [11]:
# Casting the list to a set could also be used here
sorted(observations.scientific_name.unique()) == sorted(species.scientific_name.unique())

True

<font size="3">This indicates that all unqiue species are contained in both. Considering the fact that the vast majority of species have **NaN** for conservation status, and no species scientific names are missing in the observation data, it is safe to say that since all the animals and plants in the species data are being observed, that a missing conservation status means that the species is not under any threat.<br><br>Thus, the null values will be changed to a new category *"`not in danger`"* after some other analysis.<br><br> Null Values will be treated as a species not under conservation watch.</font>

In [12]:
# species.conservation_status.isna().sum()

In [13]:
# species.fillna('`not in danger`', inplace=True)
# species

In [14]:
species.conservation_status.value_counts().sum()

191

<font size="3">From looking at the value count, I know there are only 191 species have some kind of conservation status. For the sake of this analysis, I will consider this group with some negative conservation status as *`Under Watch`* or *`In Danger`*. This will be how I refer to these species.<br><br>
Let's look at the percentage of each species under watch based on their Class (or category in the data)</font>

In [15]:
cat_counts =species.category.value_counts().to_dict()

In [16]:
cat_counts =species.category.value_counts().to_dict()
under_watch = species.groupby(['category', 'conservation_status'])['scientific_name'].count().reset_index()
# could use a lambda function here
def percentages(row):
    return row['scientific_name']/cat_counts[row['category']]
under_watch['percentage'] = under_watch.apply(percentages, axis=1)
under_watch.sort_values(by='percentage', ascending=False)

category conservation_status  scientific_name  percentage
5                Bird  Species of Concern               72    0.138196
11             Mammal  Species of Concern               28    0.130841
14            Reptile  Species of Concern                5    0.063291
1           Amphibian  Species of Concern                4    0.050000
9              Mammal          Endangered                7    0.032710
7                Fish  Species of Concern                4    0.031496
8                Fish          Threatened                4    0.031496
2           Amphibian          Threatened                2    0.025000
6                Fish          Endangered                3    0.023622
13  Nonvascular Plant  Species of Concern                5    0.015015
0           Amphibian          Endangered                1    0.012500
16     Vascular Plant  Species of Concern               43    0.009620
12             Mammal          Threatened                2    0.009346
3                Bird          Endangered                4    0.007678
4                Bird         In Recovery                3    0.005758
10             Mammal         In Recovery                1    0.004673
17     Vascular Plant          Threatened                2    0.000447
15     Vascular Plant          Endangered                1    0.000224

<font size="3">It is clear from the chart that most species are `not in danger`. Of those that are, birds and mammals have about 13% as `species of concern`. Other statuses and classes of species are at 6% total or less. The other numbers are low, especially with plants where only around 1.5% of species are of concern or worse.<br><br>Let's move on to look at the duplicated rows.
</font>

In [17]:
species.duplicated().sum()

0

<font size="3">There are no complete duplicate rows indicating that any duplication is coming from the scientific name or common name only, as per our counts earlier. I suspect there are different common names for some species and some scientific names that are duplicated. 
<br><br>
Let's take a look at how many there are of each.</font>

In [18]:
species.duplicated(subset=['scientific_name']).sum(), species.duplicated(subset=['common_names']).sum()

(283, 320)

<font size="3">Let's see which share both scientific and common names</font>

In [19]:
species[species.duplicated(subset=['scientific_name','common_names'], keep=False)].sort_values(by='common_names')

category      scientific_name     common_names conservation_status
3020   Mammal          Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal          Canis lupus  Gray Wolf, Wolf          Endangered
560      Fish  Oncorhynchus mykiss    Rainbow Trout                 NaN
3283     Fish  Oncorhynchus mykiss    Rainbow Trout          Threatened

<font size="3">Only two species, Canis lupus and Oncorhynchus mykiss, which I will refer in the future by the common names, have shared scientific and common names. In addition, they also have mixed statuses. I will come back and look in more detail at these two particular species<br><br>
Let's look next at species with duplicated common names first to see their conservation status.</font>

In [20]:
duplicated_common_names = species[species.duplicated(subset=['common_names'], keep=False)]
duplicated_common_names.sort_values(by='common_names')#[~duplicated_common_names.conservation_status.isna()]

category                      scientific_name  \
2730  Nonvascular Plant              Dichodontium pellucidum   
2822  Nonvascular Plant               Cirriphyllum piliferum   
2022     Vascular Plant                       Carex normalis   
1971     Vascular Plant                      Carex bromoides   
1960     Vascular Plant                      Carex annectens   
...                 ...                                  ...   
250                Bird                   Dendroica coronata   
252                Bird                   Dendroica dominica   
3206               Bird                   Setophaga dominica   
2957  Nonvascular Plant  Zygodon viridissimus var. rupestris   
2956  Nonvascular Plant                 Zygodon viridissimus   

                 common_names conservation_status  
2730                   A Moss                 NaN  
2822                   A Moss                 NaN  
2022           A Sedge, Sedge                 NaN  
1971           A Sedge, Sedge                 NaN  
1960           A Sedge, Sedge                 NaN  
...                       ...                 ...  
250     Yellow-Rumped Warbler                 NaN  
252   Yellow-Throated Warbler                 NaN  
3206  Yellow-Throated Warbler                 NaN  
2957             Zygodon Moss                 NaN  
2956             Zygodon Moss                 NaN  

[568 rows x 4 columns]

In [21]:
duplicated_common_names.conservation_status.value_counts(dropna=False)

NaN                   556
Species of Concern      9
In Recovery             1
Threatened              1
Endangered              1
Name: conservation_status, dtype: int64

<font size="3">The majority of the duplicated species are `not in danger`. Below, I will create a list of those species who are under watch and investigate if any have mixed statuses like the gray wolf or rainbow trout.</font>

In [22]:
display(duplicated_common_names[~duplicated_common_names.conservation_status.isna()].sort_values(by='common_names'))
duplicated_common_names[~duplicated_common_names.conservation_status.isna()].sort_values(by='common_names').shape[0]

category                      scientific_name  \
2929  Nonvascular Plant                  Bazzania nudicaulis   
185                Bird                     Guiraca caerulea   
2292     Vascular Plant                       Poa paludigena   
3102               Bird                    Stellula calliope   
3020             Mammal                          Canis lupus   
4448             Mammal                          Canis lupus   
3288               Fish                       Cottus bairdii   
284                Bird                Vermivora ruficapilla   
310                Bird                     Contopus cooperi   
3283               Fish                  Oncorhynchus mykiss   
4879     Vascular Plant  Plagiobothrys torreyi var. diffusus   
287                Bird                     Wilsonia pusilla   

                common_names conservation_status  
2929                Bazzania  Species of Concern  
185            Blue Grosbeak  Species of Concern  
2292           Bog Bluegrass  Species of Concern  
3102    Calliope Hummingbird  Species of Concern  
3020         Gray Wolf, Wolf         In Recovery  
4448         Gray Wolf, Wolf          Endangered  
3288         Mottled Sculpin  Species of Concern  
284        Nashville Warbler  Species of Concern  
310   Olive-Sided Flycatcher  Species of Concern  
3283           Rainbow Trout          Threatened  
4879  Torrey's Popcornflower  Species of Concern  
287         Wilson's Warbler  Species of Concern

12

<font size="3">Only the two animals from before have statuses that are not `species of concern` and have duplicated common name. They also share the same species name. I can use this information to confirm that the other dupclicated common name species have mixed status with _``not in danger``_. <br><br>The gray wolf is included twice, so there should be 9 of them.</font>

In [23]:
mask = duplicated_common_names.conservation_status == 'Species of Concern'
spec_of_concern = duplicated_common_names[mask].common_names.to_list()
display(len(spec_of_concern))
duplicated_common_names[duplicated_common_names.common_names.isin(spec_of_concern)].sort_values(by='common_names')

9

category                      scientific_name  \
2930  Nonvascular Plant                  Bazzania tricrenata   
2929  Nonvascular Plant                  Bazzania nudicaulis   
2928  Nonvascular Plant                    Bazzania denudata   
4525               Bird                   Passerina caerulea   
185                Bird                     Guiraca caerulea   
2292     Vascular Plant                       Poa paludigena   
5628     Vascular Plant         Poa leptocoma ssp. leptocoma   
3102               Bird                    Stellula calliope   
4506               Bird                 Selasphorus calliope   
563                Fish                        Cottus bairdi   
3288               Fish                       Cottus bairdii   
3199               Bird              Oreothlypis ruficapilla   
284                Bird                Vermivora ruficapilla   
309                Bird                    Contopus borealis   
310                Bird                     Contopus cooperi   
4880     Vascular Plant   Plagiobothrys torreyi var. torreyi   
4879     Vascular Plant  Plagiobothrys torreyi var. diffusus   
287                Bird                     Wilsonia pusilla   
3196               Bird                   Cardellina pusilla   

                common_names conservation_status  
2930                Bazzania                 NaN  
2929                Bazzania  Species of Concern  
2928                Bazzania                 NaN  
4525           Blue Grosbeak                 NaN  
185            Blue Grosbeak  Species of Concern  
2292           Bog Bluegrass  Species of Concern  
5628           Bog Bluegrass                 NaN  
3102    Calliope Hummingbird  Species of Concern  
4506    Calliope Hummingbird                 NaN  
563          Mottled Sculpin                 NaN  
3288         Mottled Sculpin  Species of Concern  
3199       Nashville Warbler                 NaN  
284        Nashville Warbler  Species of Concern  
309   Olive-Sided Flycatcher                 NaN  
310   Olive-Sided Flycatcher  Species of Concern  
4880  Torrey's Popcornflower                 NaN  
4879  Torrey's Popcornflower  Species of Concern  
287         Wilson's Warbler  Species of Concern  
3196        Wilson's Warbler                 NaN

<font size="3">The numbers add up. My initial suspicion appears to be the case. Some of these animals or plants with different common names are variations on a species. The ones that are both ``not in danger`` and `species of concern` are technically different species. Thus, for all extent and purposes, they are separate entities with there own unqiue scientific name and conservation status. Furthermore, the majority are `not in danger`, while 11 of them have different converstation statuses. All will be kept as seperate entities since the scientific names are different and none of them will be dropped from the data.<br><br>
Next, let's look at the duplicated scientific names.</font>

In [24]:
duplicated_scientific_names = species[species.duplicated(subset=['scientific_name'], keep=False)].sort_values(by='scientific_name')
duplicated_scientific_names

category      scientific_name  \
5553  Vascular Plant  Agrostis capillaris   
2132  Vascular Plant  Agrostis capillaris   
2134  Vascular Plant    Agrostis gigantea   
5554  Vascular Plant    Agrostis gigantea   
4178  Vascular Plant   Agrostis mertensii   
...              ...                  ...   
5643  Vascular Plant        Vulpia myuros   
2331  Vascular Plant     Vulpia octoflora   
4290  Vascular Plant     Vulpia octoflora   
3347  Vascular Plant         Zizia aptera   
632   Vascular Plant         Zizia aptera   

                                           common_names conservation_status  
5553                  Colonial Bent, Colonial Bentgrass                 NaN  
2132                                  Rhode Island Bent                 NaN  
2134                                             Redtop                 NaN  
5554                Black Bent, Redtop, Water Bentgrass                 NaN  
4178               Arctic Bentgrass, Northern Bentgrass                 NaN  
...                                                 ...                 ...  
5643  Foxtail Fescue, Rattail Fescue, Rat-Tail Fescu...                 NaN  
2331                                      Annual Fescue                 NaN  
4290  Eight-Flower Six-Weeks Grass, Pullout Grass, S...                 NaN  
3347  Heartleaf Alexanders, Heart-Leaf Alexanders, M...                 NaN  
632                                   Golden Alexanders                 NaN  

[557 rows x 4 columns]

In [25]:
display(duplicated_scientific_names[['conservation_status']].value_counts(dropna=False))
duplicated_scientific_names[['conservation_status']].value_counts().sum()

conservation_status
NaN                    534
Species of Concern      19
Endangered               2
In Recovery              1
Threatened               1
dtype: int64

23

<font size="3">Some species with duplicated scientific names are in the database twice because they have different common names with some having more than one common name with another. Like with common names, a majority of them are `not in danger`. It is possible that these should be considered as different entities since data collection might be occurring on different groups of these species or in different locations, but I am  not sure yet. <br><br>
Let's focus on that number 23. To start, I look at the conservation status of these species and counts to see if there are any that have more than 3 entries.</font>

In [26]:
display(duplicated_scientific_names.groupby(['conservation_status', 'scientific_name'])['common_names'].count().sum())
display(duplicated_scientific_names.groupby(['conservation_status', 'scientific_name'])['common_names'].count())
print("\nValue Counts")
duplicated_scientific_names.scientific_name.value_counts().value_counts()

23

conservation_status  scientific_name          
Endangered           Canis lupus                  2
In Recovery          Canis lupus                  1
Species of Concern   Eptesicus fuscus             2
                     Gavia immer                  2
                     Lasionycteris noctivagans    2
                     Myotis californicus          2
                     Myotis lucifugus             3
                     Nycticorax nycticorax        2
                     Pandion haliaetus            2
                     Riparia riparia              2
                     Taxidea taxus                2
Threatened           Oncorhynchus mykiss          1
Name: common_names, dtype: int64


Value Counts


2    265
3      9
Name: scientific_name, dtype: int64

In [27]:
duplicated_scientific_names.groupby(['conservation_status', 'scientific_name'])['common_names'].count().index.get_level_values('scientific_name').nunique()

11

<font size="3">All duplicated species by scientific name only have 2 or 3 duplicate entries. The number of species under watch is 11. Only 9 species total have 3 duplicated scientific names whereas 265 have 2 duplicated names. It appears that the Canis lupus (the gray wolf) and Myotis lucifugus (the little brown bat) both have three statuses under watch, with the little brown bat having all three as as `species of concern`. I already know the gray wolf and rainbow trout have mixed statuses. That leaves us with the remaining 9 of 11. So what about the duplicated species with 2 entries? Are any of them 1 of the 7 to remaning on the 3 species list? <br><br>
Let's look more at the species on this list to see which with a count of 2 have a mixed statuses and confirm the statuses of species with 3 entries.</font>

In [28]:
list_of_3_entries = duplicated_scientific_names.scientific_name.value_counts()[duplicated_scientific_names.scientific_name.value_counts() == 3].index
list_of_2_entries = duplicated_scientific_names.scientific_name.value_counts()[duplicated_scientific_names.scientific_name.value_counts() == 2].index

In [29]:
duplicated_scientific_names[duplicated_scientific_names.scientific_name.isin(list_of_3_entries)]

category        scientific_name  \
8             Mammal            Canis lupus   
3020          Mammal            Canis lupus   
4448          Mammal            Canis lupus   
49            Mammal      Castor canadensis   
4475          Mammal      Castor canadensis   
3050          Mammal      Castor canadensis   
4513            Bird          Columba livia   
3138            Bird          Columba livia   
156             Bird          Columba livia   
4236  Vascular Plant         Holcus lanatus   
5600  Vascular Plant         Holcus lanatus   
2242  Vascular Plant         Holcus lanatus   
3488  Vascular Plant   Hypochaeris radicata   
4799  Vascular Plant   Hypochaeris radicata   
858   Vascular Plant   Hypochaeris radicata   
4467          Mammal       Myotis lucifugus   
3042          Mammal       Myotis lucifugus   
37            Mammal       Myotis lucifugus   
3030          Mammal          Procyon lotor   
26            Mammal          Procyon lotor   
4459          Mammal          Procyon lotor   
16            Mammal          Puma concolor   
4451          Mammal          Puma concolor   
3022          Mammal          Puma concolor   
4514            Bird  Streptopelia decaocto   
3140            Bird  Streptopelia decaocto   
3077            Bird  Streptopelia decaocto   

                                           common_names conservation_status  
8                                             Gray Wolf          Endangered  
3020                                    Gray Wolf, Wolf         In Recovery  
4448                                    Gray Wolf, Wolf          Endangered  
49                              American Beaver, Beaver                 NaN  
4475                                             Beaver                 NaN  
3050                                    American Beaver                 NaN  
4513                                        Rock Pigeon                 NaN  
3138              Common Pigeon, Rock Dove, Rock Pigeon                 NaN  
156                                           Rock Dove                 NaN  
4236                  Common Velvetgrass, Yorkshire-Fog                 NaN  
5600     Common Velvetgrass, Velvetgrass, Yorkshire Fog                 NaN  
2242                   Common Velvet Grass, Velvetgrass                 NaN  
3488          Spotted Cats-Ear, Hairy Cats-Ear, Gosmore                 NaN  
4799  Common Cat's-Ear, False Dandelion, Frogbit, Go...                 NaN  
858                        Cat's Ear, Spotted Cat's-Ear                 NaN  
4467                                Little Brown Myotis  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern  
3030          Common Raccoon, Northern Raccoon, Raccoon                 NaN  
26                              Common Raccoon, Raccoon                 NaN  
4459                                            Raccoon                 NaN  
16                              Panther (Mountain Lion)                 NaN  
4451                                      Mountain Lion                 NaN  
3022                        Cougar, Mountain Lion, Puma                 NaN  
4514                             Eurasian Collared Dove                 NaN  
3140     Eurasian Collared Dove, Eurasian Collared-Dove                 NaN  
3077                             Eurasian Collared-Dove                 NaN

<font size="3"> It is confirmed that none of the list of 3 except the wolf and bat have mixed statuses. Therefore, the remaining 9 species (11 including ranibow trout, gray wolf and brown bat) must have only 2 entries each.<br><br>Since the list of 2 is large (265), let's create a mask to filter out any species that have both statuses as _`not in danger`_. I will test the mask first with the list of 3.</font>

In [30]:
mask_3 = ~(duplicated_scientific_names.conservation_status.isna()) & duplicated_scientific_names.scientific_name.isin(list_of_3_entries)
duplicated_scientific_names[mask_3]#.scientific_name.nunique()

category   scientific_name  \
8      Mammal       Canis lupus   
3020   Mammal       Canis lupus   
4448   Mammal       Canis lupus   
4467   Mammal  Myotis lucifugus   
3042   Mammal  Myotis lucifugus   
37     Mammal  Myotis lucifugus   

                                           common_names conservation_status  
8                                             Gray Wolf          Endangered  
3020                                    Gray Wolf, Wolf         In Recovery  
4448                                    Gray Wolf, Wolf          Endangered  
4467                                Little Brown Myotis  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
37                Little Brown Bat, Little Brown Myotis  Species of Concern

In [31]:
mask_2 = ~(duplicated_scientific_names.conservation_status.isna()) & duplicated_scientific_names.scientific_name.isin(list_of_2_entries)
duplicated_scientific_names[mask_2]

category            scientific_name  \
29     Mammal           Eptesicus fuscus   
3035   Mammal           Eptesicus fuscus   
3150     Bird                Gavia immer   
172      Bird                Gavia immer   
30     Mammal  Lasionycteris noctivagans   
3037   Mammal  Lasionycteris noctivagans   
4465   Mammal        Myotis californicus   
3039   Mammal        Myotis californicus   
337      Bird      Nycticorax nycticorax   
4564     Bird      Nycticorax nycticorax   
3283     Fish        Oncorhynchus mykiss   
3081     Bird          Pandion haliaetus   
104      Bird          Pandion haliaetus   
226      Bird            Riparia riparia   
3185     Bird            Riparia riparia   
3029   Mammal              Taxidea taxus   
4457   Mammal              Taxidea taxus   

                                           common_names conservation_status  
29                                        Big Brown Bat  Species of Concern  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
172                                         Common Loon  Species of Concern  
30                                    Silver-Haired Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
4465                                  California Myotis  Species of Concern  
3039  California Myotis, California Myotis, Californ...  Species of Concern  
337                           Black-Crowned Night-Heron  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern  
3283                                      Rainbow Trout          Threatened  
3081                             Osprey, Western Osprey  Species of Concern  
104                                              Osprey  Species of Concern  
226                                        Bank Swallow  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
3029                            American Badger, Badger  Species of Concern  
4457                                             Badger  Species of Concern

<font size="3">Because there are two entries for each `species of concern` in this filtered data, none of the entries under watch have a mixed status, except for the rainbow trout. That means the majority of entries of 2 have a status of `not in danger`.<br><br> Finally, I will use my masks to make sure all 11 species with duplicates under watch Ire accounted for.</font>

In [32]:
duplicated_scientific_names[mask_2 + mask_3].scientific_name.nunique()

# display(sorted(duplicated_scientific_names[mask].scientific_name.unique()))
# sorted(duplicated_scientific_names.groupby(['conservation_status', 'scientific_name'])['common_names'].count().to_frame().index.get_level_values(1).unique()[1:])

11

<font size="3">In summary: <ul><li>All duplicated species are either on the list 2 or 3 times.</li><li>Almost all of the species with multiple entries have the same conservation status. The majority here are _`not in danger`_</li><li>**11** species are _in danger_ or _under watch_.</li><li> **9 of 11** have a status of 'species of concern. Of those 9, *only the little brown bat has 3 duplicates* while the rest have 2.</li><li>Only **2 of 11** species, *the gray wolf and rainbow trout*, have multiple or mixed statuses.</li></ul><br>
I conclude that the duplicate entries are most likely not by mistake. They might be included more than once because of a previous change in status. The authorities are constantly updating the conservation status of species. Alternatively, it is possible that there Ire different pools or populations of the species being monitored in different locations, or they Ire being monitored by different observers. This might explain the different common names listed. Since there is no date data, it is not possible to make a definitive conclusion.
<br><br>
It is crucial to look at the obervations table to get a sense of what is going with the multiple entries.</font>

<font size="3">Before that, I am going to save the data frame containing all the duplicate species with mixed status</font>

In [33]:
duplicated_scientific_names_under_watch = duplicated_scientific_names[~duplicated_scientific_names.conservation_status.isna()]['scientific_name'].unique().tolist()
in_danger = duplicated_scientific_names[duplicated_scientific_names.scientific_name.apply(lambda species: species in duplicated_scientific_names_under_watch)].reset_index()
in_danger.to_csv('data/species_in_danger.csv')

#### Observation Data Frame Analysis

In [34]:
display(observations.shape)
observations.head()

(23296, 3)

scientific_name                            park_name  observations
0        Vicia benghalensis  Great Smoky Mountains National Park            68
1            Neovison vison  Great Smoky Mountains National Park            77
2         Prunus subcordata               Yosemite National Park           138
3      Abutilon theophrasti                  Bryce National Park            84
4  Githopsis specularioides  Great Smoky Mountains National Park            85

<font size="3">The observations data contains only the scientific name, park name, and the number of observations. There is no date or time data included, nor data about the common name. Also, it is unclear if an observation is a unique count of a species or just a sighting of a species, but I assume it is the case that an observation is a unqiue number of the species found in the park at a given time.</font>

In [35]:
observations.duplicated().sum()

15

In [36]:
observations[observations.duplicated(keep=False)].sort_values(by='scientific_name')#.head()

scientific_name                            park_name  \
513             Arctium minus               Yosemite National Park   
10674           Arctium minus               Yosemite National Park   
4527   Botrychium virginianum            Yellowstone National Park   
20294  Botrychium virginianum            Yellowstone National Park   
19392       Cichorium intybus            Yellowstone National Park   
14142       Cichorium intybus            Yellowstone National Park   
7263   Echinochloa crus-galli  Great Smoky Mountains National Park   
1454   Echinochloa crus-galli  Great Smoky Mountains National Park   
1020     Eleocharis palustris  Great Smoky Mountains National Park   
12381    Eleocharis palustris  Great Smoky Mountains National Park   
1032      Hesperis matronalis                  Bryce National Park   
13372     Hesperis matronalis                  Bryce National Park   
4178     Monotropa hypopithys  Great Smoky Mountains National Park   
20820    Monotropa hypopithys  Great Smoky Mountains National Park   
10038          Plantago major  Great Smoky Mountains National Park   
365            Plantago major  Great Smoky Mountains National Park   
18141           Poa compressa  Great Smoky Mountains National Park   
13097           Poa compressa  Great Smoky Mountains National Park   
17378    Potentilla norvegica               Yosemite National Park   
10809    Potentilla norvegica               Yosemite National Park   
7219        Prunella vulgaris  Great Smoky Mountains National Park   
16741       Prunella vulgaris  Great Smoky Mountains National Park   
19575            Salix exigua               Yosemite National Park   
22698            Salix exigua               Yosemite National Park   
17577      Spilogale putorius            Yellowstone National Park   
627        Spilogale putorius            Yellowstone National Park   
8710    Streptopelia decaocto            Yellowstone National Park   
1635    Streptopelia decaocto            Yellowstone National Park   
21151     Trifolium campestre            Yellowstone National Park   
18435     Trifolium campestre            Yellowstone National Park   

       observations  
513             162  
10674           162  
4527            232  
20294           232  
19392           266  
14142           266  
7263             62  
1454             62  
1020             62  
12381            62  
1032            124  
13372           124  
4178             73  
20820            73  
10038            90  
365              90  
18141            80  
13097            80  
17378           148  
10809           148  
7219             75  
16741            75  
19575           186  
22698           186  
17577           256  
627             256  
8710            255  
1635            255  
21151           239  
18435           239

<font size="3">Given the large number of observations, it is safe to say that duplicates are just measures of a species with the same number taken at different times or by different observers. It would mean that the population numbers is stable for that species in that park over some time period, or that it fluctuated and returned to a previous population number<br><br>Next, I want to look at some aggregated data to confirm a suspicion: **each species is measured once at each park** and **four times in total**.</font>

In [37]:
species.shape[0] * 4 == observations.shape[0]

True

<font size="3">The number in the obserations data set is 4 times the number in the species data set. This makes sense that each speicies from the data set is observed in the four different parks, assumingly at the same time frame.<br><br>This will be confirmed with a check below.</font>f

In [38]:
sorted_species_park = observations.sort_values(by=['scientific_name', 'park_name'])
display(sorted_species_park.tail(12))
sorted_species_park.head(12)

scientific_name  \
5350                   Zonotrichia querula   
15795                  Zonotrichia querula   
17507                  Zonotrichia querula   
12923                  Zonotrichia querula   
16130                 Zygodon viridissimus   
3649                  Zygodon viridissimus   
5015                  Zygodon viridissimus   
5539                  Zygodon viridissimus   
12775  Zygodon viridissimus var. rupestris   
20040  Zygodon viridissimus var. rupestris   
6879   Zygodon viridissimus var. rupestris   
16237  Zygodon viridissimus var. rupestris   

                                 park_name  observations  
5350                   Bryce National Park           105  
15795  Great Smoky Mountains National Park            83  
17507            Yellowstone National Park           268  
12923               Yosemite National Park           160  
16130                  Bryce National Park           100  
3649   Great Smoky Mountains National Park            71  
5015             Yellowstone National Park           270  
5539                Yosemite National Park           159  
12775                  Bryce National Park           102  
20040  Great Smoky Mountains National Park           102  
6879             Yellowstone National Park           237  
16237               Yosemite National Park           210

scientific_name                            park_name  observations
4968    Abies bifolia                  Bryce National Park           109
17990   Abies bifolia  Great Smoky Mountains National Park            72
8050    Abies bifolia            Yellowstone National Park           215
20930   Abies bifolia               Yosemite National Park           136
8170   Abies concolor                  Bryce National Park            83
4647   Abies concolor  Great Smoky Mountains National Park           101
15193  Abies concolor            Yellowstone National Park           241
23166  Abies concolor               Yosemite National Park           205
2014    Abies fraseri                  Bryce National Park           109
5112    Abies fraseri  Great Smoky Mountains National Park            81
11141   Abies fraseri            Yellowstone National Park           218
5793    Abies fraseri               Yosemite National Park           110

<font size="3">This also implies that the duplicates in the species with scientific names data are not by mistake, but are seperate sets of observation measures.<br><br>Let's do some confirmation of this by looking at the value counts of value counts. *I know from are dupcliate analyis that the duplicates Ire in groups of 3 or 2*, so the observations should only have values of 4, 8 and 12.</font>

In [39]:
obsv_counts = observations.scientific_name.value_counts()
display(obsv_counts)
display(obsv_counts.to_frame().value_counts())
display(sorted_species_park[['scientific_name', 'park_name']].value_counts().value_counts())
all(obsv_counts[obsv_counts.values > 4].index.sort_values() == duplicated_scientific_names.scientific_name.unique())

Myotis lucifugus                        12
Puma concolor                           12
Hypochaeris radicata                    12
Holcus lanatus                          12
Streptopelia decaocto                   12
                                        ..
Packera dimorphophylla var. paysonii     4
Smilax bona-nox                          4
Chondestes grammacus                     4
Leymus triticoides                       4
Dichanthelium depauperatum               4
Name: scientific_name, Length: 5541, dtype: int64

scientific_name
4                  5267
8                   265
12                    9
dtype: int64

1    21068
2     1060
3       36
dtype: int64

True

<font size="3">I confirmed the names are the same and the counts match (9 species had duplicates of 3 and 265 had duplicates of 2). Sorting by name and parks also confirms that there are four observations for all species.<br><br>Therefore, I should keep all the data in the two data sets. This will cause some problems for merging the data set, though.</font>

<font size="3">Let's look at the observation data to get an idea of the meaning of the different conservation statuses. I can filter data for the duplicates using the observation counts data frame.</font>

In [40]:
display(observations[observations.scientific_name.isin(obsv_counts[:9].index)].reset_index().sort_values(by=['scientific_name', 'index']).head(12))
observations[observations.scientific_name.isin(obsv_counts[:9].index)].reset_index().sort_values(by=['scientific_name', 'index']).tail(24)

index scientific_name                            park_name  observations
4     1294     Canis lupus               Yosemite National Park            35
9     1766     Canis lupus                  Bryce National Park            27
35    7346     Canis lupus                  Bryce National Park            29
53    9884     Canis lupus                  Bryce National Park            74
55   10190     Canis lupus  Great Smoky Mountains National Park            15
56   10268     Canis lupus            Yellowstone National Park            60
63   10907     Canis lupus            Yellowstone National Park            67
72   13427     Canis lupus            Yellowstone National Park           203
92   17756     Canis lupus  Great Smoky Mountains National Park            14
98   19330     Canis lupus               Yosemite National Park           117
101  19987     Canis lupus               Yosemite National Park            44
103  20353     Canis lupus  Great Smoky Mountains National Park            30

index        scientific_name                            park_name  \
0      77          Puma concolor                  Bryce National Park   
14   2649          Puma concolor  Great Smoky Mountains National Park   
18   3270          Puma concolor            Yellowstone National Park   
31   6660          Puma concolor  Great Smoky Mountains National Park   
46   8976          Puma concolor               Yosemite National Park   
47   9298          Puma concolor               Yosemite National Park   
49   9605          Puma concolor            Yellowstone National Park   
58  10388          Puma concolor                  Bryce National Park   
81  15676          Puma concolor            Yellowstone National Park   
88  16551          Puma concolor  Great Smoky Mountains National Park   
91  17631          Puma concolor               Yosemite National Park   
95  18321          Puma concolor                  Bryce National Park   
8    1635  Streptopelia decaocto            Yellowstone National Park   
16   3200  Streptopelia decaocto                  Bryce National Park   
19   3376  Streptopelia decaocto               Yosemite National Park   
23   4515  Streptopelia decaocto                  Bryce National Park   
33   7057  Streptopelia decaocto  Great Smoky Mountains National Park   
39   8072  Streptopelia decaocto                  Bryce National Park   
44   8710  Streptopelia decaocto            Yellowstone National Park   
54  10107  Streptopelia decaocto  Great Smoky Mountains National Park   
60  10643  Streptopelia decaocto            Yellowstone National Park   
79  14699  Streptopelia decaocto  Great Smoky Mountains National Park   
80  14942  Streptopelia decaocto               Yosemite National Park   
96  18854  Streptopelia decaocto               Yosemite National Park   

    observations  
0             92  
14            79  
18           243  
31            48  
46           121  
47           165  
49           240  
58           128  
81           270  
88           112  
91           122  
95            91  
8            255  
16            92  
19           124  
23            88  
33            74  
39           121  
44           255  
54            72  
60           261  
79           110  
80           183  
96           150

<font size="3">A quick look shows that there is not particular order to the data that would help us seperate the multiple labelled species. That is, it is impossible to tell which set of observations match with the species with unique common names and the same scientific name. I kept the original indexing to see if that would maintain some sort of order in the parks, which could represent observations being made on certain days in a certain order. I was hoping to see if I could reorder or make conclusions about these multilisted species.<br><br>This is not such a problem since most of this species maintain the same status. The exception is when I am  dealing with species that had multiple statuses, the rainbow trout and gray wolf. The rest of the duplicates have the same labelled conservation status<br><br>Next, I am going to focus on the species under watch from my duplicated data frame to try to get a better sense of the differences between conversvation statuses.</font>

In [41]:
under_watch =observations[observations.scientific_name.apply(lambda species: species in duplicated_scientific_names_under_watch)].sort_values(by=['scientific_name', 'park_name'])

In [42]:
in_danger.tail()

index category    scientific_name               common_names  \
19    104     Bird  Pandion haliaetus                     Osprey   
20    226     Bird    Riparia riparia               Bank Swallow   
21   3185     Bird    Riparia riparia  Bank Swallow, Sand Martin   
22   3029   Mammal      Taxidea taxus    American Badger, Badger   
23   4457   Mammal      Taxidea taxus                     Badger   

   conservation_status  
19  Species of Concern  
20  Species of Concern  
21  Species of Concern  
22  Species of Concern  
23  Species of Concern

<font size="3">Both the species here, the badger, bank swallow and osprey, have a conservation status of `species of concern`. I have sorted the observation data by park name to glean some insight.</font>

In [43]:
under_watch.tail(24)

scientific_name                            park_name  observations
2395   Pandion haliaetus                  Bryce National Park            72
7153   Pandion haliaetus                  Bryce National Park            82
2389   Pandion haliaetus  Great Smoky Mountains National Park            58
15281  Pandion haliaetus  Great Smoky Mountains National Park            80
2585   Pandion haliaetus            Yellowstone National Park           223
5514   Pandion haliaetus            Yellowstone National Park           243
200    Pandion haliaetus               Yosemite National Park           134
3526   Pandion haliaetus               Yosemite National Park           148
872      Riparia riparia                  Bryce National Park            88
1259     Riparia riparia                  Bryce National Park            91
2672     Riparia riparia  Great Smoky Mountains National Park            92
4916     Riparia riparia  Great Smoky Mountains National Park            76
220      Riparia riparia            Yellowstone National Park           199
14216    Riparia riparia            Yellowstone National Park           226
7545     Riparia riparia               Yosemite National Park           125
18275    Riparia riparia               Yosemite National Park           145
1581       Taxidea taxus                  Bryce National Park            94
11772      Taxidea taxus                  Bryce National Park            65
5869       Taxidea taxus  Great Smoky Mountains National Park            53
18769      Taxidea taxus  Great Smoky Mountains National Park            84
11488      Taxidea taxus            Yellowstone National Park           261
20400      Taxidea taxus            Yellowstone National Park           188
3983       Taxidea taxus               Yosemite National Park           133
4708       Taxidea taxus               Yosemite National Park           122

<font size="3">For the three species, the data follows the same pattern. It seems that in two of the parks, Bryce and Great Smoky Mountains, the number of observations is under 100. The numbers for the other two parks, Yosemite and Yellowstone, are over 100 with a few over 200.<br><br>Let's compare these numbers to the gray wolf.</font>

In [44]:
in_danger.head(3)

index category scientific_name     common_names conservation_status
0      8   Mammal     Canis lupus        Gray Wolf          Endangered
1   3020   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
2   4448   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

In [45]:
under_watch.head(12)

scientific_name                            park_name  observations
1766      Canis lupus                  Bryce National Park            27
7346      Canis lupus                  Bryce National Park            29
9884      Canis lupus                  Bryce National Park            74
10190     Canis lupus  Great Smoky Mountains National Park            15
17756     Canis lupus  Great Smoky Mountains National Park            14
20353     Canis lupus  Great Smoky Mountains National Park            30
10268     Canis lupus            Yellowstone National Park            60
10907     Canis lupus            Yellowstone National Park            67
13427     Canis lupus            Yellowstone National Park           203
1294      Canis lupus               Yosemite National Park            35
19330     Canis lupus               Yosemite National Park           117
19987     Canis lupus               Yosemite National Park            44

<font size="3"> It appears that to be `endangered`, the number of observations under 100 have to appear in 3 or more parks. Once 2 of the 4 parks have numbers over 100, the species appear to be `in recovery`.</font>

In [46]:
endangered_species_names = species[species.conservation_status == 'Endangered'].scientific_name.to_list()
endangered_species = observations[observations.scientific_name.isin(endangered_species_names)].sort_values(by=['scientific_name', 'park_name'])
endangered_species.shape

(68, 3)

In [47]:
display(endangered_species.head(50))
endangered_species.tail(18)

scientific_name                            park_name  \
1766                    Canis lupus                  Bryce National Park   
7346                    Canis lupus                  Bryce National Park   
9884                    Canis lupus                  Bryce National Park   
10190                   Canis lupus  Great Smoky Mountains National Park   
17756                   Canis lupus  Great Smoky Mountains National Park   
20353                   Canis lupus  Great Smoky Mountains National Park   
10268                   Canis lupus            Yellowstone National Park   
10907                   Canis lupus            Yellowstone National Park   
13427                   Canis lupus            Yellowstone National Park   
1294                    Canis lupus               Yosemite National Park   
19330                   Canis lupus               Yosemite National Park   
19987                   Canis lupus               Yosemite National Park   
939                     Canis rufus                  Bryce National Park   
15906                   Canis rufus  Great Smoky Mountains National Park   
15940                   Canis rufus            Yellowstone National Park   
3855                    Canis rufus               Yosemite National Park   
3935              Chasmistes liorus                  Bryce National Park   
8411              Chasmistes liorus  Great Smoky Mountains National Park   
6418              Chasmistes liorus            Yellowstone National Park   
20484             Chasmistes liorus               Yosemite National Park   
19337          Etheostoma percnurum                  Bryce National Park   
9096           Etheostoma percnurum  Great Smoky Mountains National Park   
1265           Etheostoma percnurum            Yellowstone National Park   
21268          Etheostoma percnurum               Yosemite National Park   
22412                 Geum radiatum                  Bryce National Park   
2715                  Geum radiatum  Great Smoky Mountains National Park   
4224                  Geum radiatum            Yellowstone National Park   
8438                  Geum radiatum               Yosemite National Park   
7622   Glaucomys sabrinus coloratus                  Bryce National Park   
8571   Glaucomys sabrinus coloratus  Great Smoky Mountains National Park   
20570  Glaucomys sabrinus coloratus            Yellowstone National Park   
9283   Glaucomys sabrinus coloratus               Yosemite National Park   
13745                Grus americana                  Bryce National Park   
9753                 Grus americana  Great Smoky Mountains National Park   
10848                Grus americana            Yellowstone National Park   
20342                Grus americana               Yosemite National Park   
7508        Gymnogyps californianus                  Bryce National Park   
15585       Gymnogyps californianus  Great Smoky Mountains National Park   
22961       Gymnogyps californianus            Yellowstone National Park   
13861       Gymnogyps californianus               Yosemite National Park   
2563              Myotis grisescens                  Bryce National Park   
12169             Myotis grisescens  Great Smoky Mountains National Park   
8460              Myotis grisescens            Yellowstone National Park   
10411             Myotis grisescens               Yosemite National Park   
10470                Myotis sodalis                  Bryce National Park   
8736                 Myotis sodalis  Great Smoky Mountains National Park   
16148                Myotis sodalis            Yellowstone National Park   
7676                 Myotis sodalis               Yosemite National Park   
6840                Noturus baileyi                  Bryce National Park   
4207                Noturus baileyi  Great Smoky Mountains National Park   

       observations  
1766             27  
7346             29  
9884             74  
10190            15  
17756            14  
20353      

scientific_name                            park_name  \
5134           Noturus baileyi            Yellowstone National Park   
16618          Noturus baileyi               Yosemite National Park   
16883  Ovis canadensis sierrae                  Bryce National Park   
17705  Ovis canadensis sierrae  Great Smoky Mountains National Park   
6761   Ovis canadensis sierrae            Yellowstone National Park   
10279  Ovis canadensis sierrae               Yosemite National Park   
20972        Picoides borealis                  Bryce National Park   
17732        Picoides borealis  Great Smoky Mountains National Park   
19139        Picoides borealis            Yellowstone National Park   
271          Picoides borealis               Yosemite National Park   
16314             Rana sierrae                  Bryce National Park   
18823             Rana sierrae  Great Smoky Mountains National Park   
11904             Rana sierrae            Yellowstone National Park   
484               Rana sierrae               Yosemite National Park   
6458       Vermivora bachmanii                  Bryce National Park   
18558      Vermivora bachmanii  Great Smoky Mountains National Park   
17418      Vermivora bachmanii            Yellowstone National Park   
13982      Vermivora bachmanii               Yosemite National Park   

       observations  
5134             67  
16618            31  
16883            22  
17705            25  
6761             67  
10279            39  
20972            24  
17732            15  
19139            64  
271              43  
16314            31  
18823            11  
11904            60  
484              42  
6458             20  
18558            18  
17418            58  
13982            45

In [48]:
threatened_species_names = species[species.conservation_status == 'Threatened'].scientific_name.to_list()
threatened_species = observations[observations.scientific_name.isin(threatened_species_names)].sort_values(by=['scientific_name', 'park_name'])
threatened_species.shape

(44, 3)

In [49]:
threatened_species.head(44)

scientific_name                            park_name  \
15049               Anaxyrus canorus                  Bryce National Park   
497                 Anaxyrus canorus  Great Smoky Mountains National Park   
20897               Anaxyrus canorus            Yellowstone National Park   
15806               Anaxyrus canorus               Yosemite National Park   
10192              Erimonax monachus                  Bryce National Park   
17068              Erimonax monachus  Great Smoky Mountains National Park   
16275              Erimonax monachus            Yellowstone National Park   
4076               Erimonax monachus               Yosemite National Park   
4989             Isotria medeoloides                  Bryce National Park   
22288            Isotria medeoloides  Great Smoky Mountains National Park   
4024             Isotria medeoloides            Yellowstone National Park   
17423            Isotria medeoloides               Yosemite National Park   
18171         Myotis septentrionalis                  Bryce National Park   
3365          Myotis septentrionalis  Great Smoky Mountains National Park   
7467          Myotis septentrionalis            Yellowstone National Park   
19485         Myotis septentrionalis               Yosemite National Park   
17679            Noturus flavipinnis                  Bryce National Park   
10991            Noturus flavipinnis  Great Smoky Mountains National Park   
10810            Noturus flavipinnis            Yellowstone National Park   
9551             Noturus flavipinnis               Yosemite National Park   
12890  Oncorhynchus clarkii henshawi                  Bryce National Park   
5919   Oncorhynchus clarkii henshawi  Great Smoky Mountains National Park   
17626  Oncorhynchus clarkii henshawi            Yellowstone National Park   
19173  Oncorhynchus clarkii henshawi               Yosemite National Park   
925              Oncorhynchus mykiss                  Bryce National Park   
11893            Oncorhynchus mykiss                  Bryce National Park   
3354             Oncorhynchus mykiss  Great Smoky Mountains National Park   
15020            Oncorhynchus mykiss  Great Smoky Mountains National Park   
4649             Oncorhynchus mykiss            Yellowstone National Park   
8682             Oncorhynchus mykiss            Yellowstone National Park   
167              Oncorhynchus mykiss               Yosemite National Park   
15239            Oncorhynchus mykiss               Yosemite National Park   
5127                  Rana draytonii                  Bryce National Park   
344                   Rana draytonii  Great Smoky Mountains National Park   
8054                  Rana draytonii            Yellowstone National Park   
470                   Rana draytonii               Yosemite National Park   
19817             Spiraea virginiana                  Bryce National Park   
8455              Spiraea virginiana  Great Smoky Mountains National Park   
16717             Spiraea virginiana            Yellowstone National Park   
5873              Spiraea virginiana               Yosemite National Park   
8750         Ursus arctos horribilis                  Bryce National Park   
14337        Ursus arctos horribilis  Great Smoky Mountains National Park   
19660        Ursus arctos horribilis            Yellowstone National Park   
18906        Ursus arctos horribilis               Yosemite National Park   

       observations  
15049            43  
497              30  
20897           114  
15806            87  
10192            46  
17068            18  
16275           109  
4076             69  
4989             42  
22288            36  
4024            114  
17423            82  
18171            52  
3365             34  
7467            143  
19485            73  
17679            55  
10991            36  
10810           126  
9551             72  
12890            57  
5919             48  
17626           111  
19173            85  
925  

<font size="3">After looking at the data more closely, I can deduce the hierarchy of conseration status:<ol><li>`Endangered` means the numbers are *less than 100* in **all four** parks</li><li>`Threatened` means the numbers are *less than 100* in **three of four** parks</li><li>`Species of Concern` means the numbers are *less than 100* in **two of four** parks</li></ol></font>

### Further Analysis

<font size="3">More analysis with visualizations will be done in a second notebook. In order to do that, I will first make two data frames of merged data. The second will contain all the raw data for the gray wolf and rainbow trout, since they are the only two species with mixed or changing statuses.The first will contain all the data connected on *scientific name* by first dropping the duplicated rows in the species column with the same scienitific name. **Even though these might be different pools of the species or different sets of observations, given the different common names, all of these species with multiple observations have the same conservation status.** Thus, I believe for the sake of analysis that we can combine the sets of data, with the exception of the rainbow trout and gray wolf.<br><br>I will perform some double-checking on the data before mergring.</font>

In [50]:
species_mask = species.scientific_name.isin(['Canis lupus', 'Oncorhynchus mykiss'])
species_2 = species[~species_mask].copy()

In [51]:
# Double Check on the duplicated species names
display(in_danger.scientific_name.unique())
vc = species_2.groupby(['scientific_name', 'conservation_status']).count()
vc[vc.category > 1]

array(['Canis lupus', 'Eptesicus fuscus', 'Gavia immer',
       'Lasionycteris noctivagans', 'Myotis californicus',
       'Myotis lucifugus', 'Nycticorax nycticorax', 'Oncorhynchus mykiss',
       'Pandion haliaetus', 'Riparia riparia', 'Taxidea taxus'],
      dtype=object)

category  common_names
scientific_name           conservation_status                        
Eptesicus fuscus          Species of Concern          2             2
Gavia immer               Species of Concern          2             2
Lasionycteris noctivagans Species of Concern          2             2
Myotis californicus       Species of Concern          2             2
Myotis lucifugus          Species of Concern          3             3
Nycticorax nycticorax     Species of Concern          2             2
Pandion haliaetus         Species of Concern          2             2
Riparia riparia           Species of Concern          2             2
Taxidea taxus             Species of Concern          2             2

<font size="3">These match up. It is safe to drop the duplicates since all other species match to only a single convservation status and these duplcated all have the same status and correct counts.</font>

In [52]:
species_2.drop_duplicates(subset='scientific_name', inplace=True)

In [53]:
obsv_mask = observations.scientific_name.isin(['Canis lupus', 'Oncorhynchus mykiss'])

observations_2 = observations[~obsv_mask]

In [54]:
display(species_2.shape[0] * 4)
observations_2.shape[0] - 264*4 - 8*8

22156

22156

<font size="3">There were 265 entries of 2 (minus the rainbow trout) and 9 entries of 3 (minus the gray). Multiplying the species numbers by 4 and comparing to the observations data where the extra data from the entries of 2 and 3 was subtracted shows the same number. Thus, it is safe to perform a SQL-like merge of the data frames.</font>

In [55]:
biodiversity = species_2.merge(observations_2, on='scientific_name', how='left')

In [56]:
biodiversity.fillna('Not in Danger', inplace=True)

In [57]:
biodiversity.to_csv('biodiversity.csv')

In [58]:
species_trout_wolf = species[species_mask].reset_index().rename(columns={'index': 'species_index'}).copy()
species_trout_wolf.fillna('Not in Danger', inplace=True)
obsv_trout_wolf = observations[obsv_mask].reset_index().rename(columns={'index': 'observation_index'}).copy()

In [59]:
obsv_trout_wolf

observation_index      scientific_name  \
0                 167  Oncorhynchus mykiss   
1                 925  Oncorhynchus mykiss   
2                1294          Canis lupus   
3                1766          Canis lupus   
4                3354  Oncorhynchus mykiss   
5                4649  Oncorhynchus mykiss   
6                7346          Canis lupus   
7                8682  Oncorhynchus mykiss   
8                9884          Canis lupus   
9               10190          Canis lupus   
10              10268          Canis lupus   
11              10907          Canis lupus   
12              11893  Oncorhynchus mykiss   
13              13427          Canis lupus   
14              15020  Oncorhynchus mykiss   
15              15239  Oncorhynchus mykiss   
16              17756          Canis lupus   
17              19330          Canis lupus   
18              19987          Canis lupus   
19              20353          Canis lupus   

                              park_name  observations  
0                Yosemite National Park           118  
1                   Bryce National Park            59  
2                Yosemite National Park            35  
3                   Bryce National Park            27  
4   Great Smoky Mountains National Park            61  
5             Yellowstone National Park           119  
6                   Bryce National Park            29  
7             Yellowstone National Park           253  
8                   Bryce National Park            74  
9   Great Smoky Mountains National Park            15  
10            Yellowstone National Park            60  
11            Yellowstone National Park            67  
12                  Bryce National Park           105  
13            Yellowstone National Park           203  
14  Great Smoky Mountains National Park            39  
15               Yosemite National Park            59  
16  Great Smoky Mountains National Park            14  
17               Yosemite National Park           117  
18               Yosemite National Park            44  
19  Great Smoky Mountains National Park            30

In [60]:
trout_wolf = species_trout_wolf.join(obsv_trout_wolf.set_index('scientific_name'), on='scientific_name', how='left')

In [61]:
trout_wolf

species_index category      scientific_name     common_names  \
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
0              8   Mammal          Canis lupus        Gray Wolf   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
1            560     Fish  Oncorhynchus mykiss    Rainbow Trout   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
2           3020   Mammal          Canis lupus  Gray Wolf, Wolf   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
3           3283     Fish  Oncorhynchus mykiss    Rainbow Trout   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   
4           4448   Mammal          Canis lupus  Gray Wolf, Wolf   

  conservation_status  observation_index                            park_name  \
0          Endangered               1294               Yosemite National Park   
0          Endangered               1766                  Bryce National Park   
0          Endangered               7346                  Bryce National Park   
0          Endangered               9884                  Bryce National Park   
0          Endangered              10190  Grea

In [62]:
trout_wolf.to_csv('data/trout_wolf.csv')

<font size="3">The total data has been joined and will be visualized in the next notebook</font>